### ISSUE TITLE
GlobalIndexUidAggregator can drop UIDs in the remove list

### ISSUE THREAD

**Issue by brianloss** [NationalSecurityAgency/datawave#1173]
GlobalIndexUidAggregator is typically driven by TotalAggregatingIterator, even though it extends Combiner and probably could be made to run independently as a Combiner. TotalAggregatingIterator appears to treat scan, minor, and non-full major compactions in the same way with special behavior for a full major compaction. If a non-full major compaction happens in the right order, then it is possible for UIDs in the removed list to be dropped (where the UID will remain in the UID list and just the removal will be dropped).

I simulated the right circumstances by starting with a table configured with TotalAggregatingIterator/GlobalIndexUidAggregator. Then, I inserted some keys and scanned to get the output that would be produced by a major compaction involving those keys. I then took that output and inserted in an empty table (to simulate the table now having the result of the non-full major compaction). I then added a new key and scanned.

```
> insert term "" "" '\x08\x00\x10\x01\x1A\x05uuid2' -ts 20
> insert term "" "" '\x08\x00\x10\x01\x1A\x05uuid3' -ts 30
> insert term "" "" '\x08\x00\x10\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\x01"\x05uuid1' -ts 40
> scan -st
term : [] 40    \x08\x00\x10\x01\x1A\x05uuid2\x1A\x05uuid3"\x05uuid1
```

Here we see the protocol buffer has uuid2, uuid3 in the UID list and uuid1 in the remove list. The total count is 1 since we had +1 each for uuid2 and uuid3 and -1 for uuid1. Now, I wiped the table and inserted two keys:

```
> insert term "" "" '\x08\x00\x10\x01\x1A\x05uuid2\x1A\x05uuid3"\x05uuid1' -ts 40
> insert term "" "" '\x08\x00\x10\x01\x1A\x05uuid1' -ts 10
> scan -st
term : [] 40    \x08\x00\x10\x02\x1A\x05uuid2\x1A\x05uuid1\x1A\x05uuid3
```

The first insert represents the result of the partial major compaction that would have happened had the files containing the first 3 inserts above been compacted together. Note that the timestamp of uuid1 (10) is older than the removal of uuid1 (40), so uuid1 should be gone. However, the final  protocol buffer has uuid1, uuid2, and uuid3 in the UID list (though the count of 2 is what it should be).

When I initially tried this test, I tried inserting just a single uuid, but found that it plus the delete yielded a count of 0 which is also not handled in the reduce method in GlobalIndexUidAggregator:

```
> insert term "" "" '\x08\x00\x10\x01\x1A\x05uuid2' -ts 20
> insert term "" "" '\x08\x00\x10\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\x01"\x05uuid1' -ts 40
> scan -st
term : [] 40    \x08\x00\x10\x00\x1A\x05uuid2"\x05uuid1
> compact -w
> scan -st
>
```

**Comment by ivakegg** [comment]
This may be covered by #1107 or visa versa.

**Comment by jschmidt10** [comment]
I found a few additional cases where we can drop removal uids.

The propogate flag gets set to false when we encounter a key that includes a protobuf w/ignore = true. The propogate flag never gets reset back to true - so all remaining keys in the compaction/scan will have propogate = false.

The propogateKey() method will return false for protobufs that only have uidsToRemove and a negative count. We want to preserve these though. 

@brianloss Mind if I take a stab at addressing this?

**Comment by brianloss** [comment]
> I found a few additional cases where we can drop removal uids.
> 
> The propogate flag gets set to false when we encounter a key that includes a protobuf w/ignore = true. The propogate flag never gets reset back to true - so all remaining keys in the compaction/scan will have propogate = false.
> 
> The propogateKey() method will return false for protobufs that only have uidsToRemove and a negative count. We want to preserve these though.
> 
> @brianloss Mind if I take a stab at addressing this?

@jschmidt10 I have some significant local changes where I was attempting to rework a little bit (based on previous comments). If it's ok, give me a day to look over those again and push.

**Comment by jschmidt10** [comment]
@brianloss Sure thing!

**Comment by brianloss** [comment]
@jschmidt10 I have pushed changes (along with targeting against the integration branch as requested). I looked at the cases you mentioned, as well. See comments below...

> I found a few additional cases where we can drop removal uids.
> 
> The propogate flag gets set to false when we encounter a key that includes a protobuf w/ignore = true. The propogate flag never gets reset back to true - so all remaining keys in the compaction/scan will have propogate = false.

I believe the propogate flag should never have been set to false in the aggregate method. Setting that to false breaks the behavior for other keys potentially, and is also what caused the need to add the check in the propogateKey method for `seenIgnore && count > maxUids`. Instead, I changed the code to leave the propogate flag alone. I updated the check in the propogateKey method and added comments for why I think what's there is correct.

> 
> The propogateKey() method will return false for protobufs that only have uidsToRemove and a negative count. We want to preserve these though.

This case should also be covered by my changes to propogateKey. If the propogate flag is false, that means the aggregated result includes all possible values for a key. Therefore, if there were a protocol buffer with only a negative count and values in uidsToRemove, we'd want to discard it
